### PR TITLE
Datasets 2.0

### PR DESCRIPTION
**Pull request by timonpalm**
I refactored the dataset classes. The manually parsing of the images folders was meticulous and there were no information about the classes and their labels. Thus, I fall back on the pytorch class, which can read the meta.bin. Otherwise, I would have just copied their implementation for that.

Further, I implemented a function to display some basic stats of the datasets (i.e distribution per class etc.). It is also now possible to create a subset of each dataset with n_samples per class. So the subsets will also always be uniformly distributed among the classes.

### CODE REVIEW

File: src/pml_vqvae/train.py [hunk 1]
```diff
@@ -1,17 +1,96 @@
...
+            output = model(batch_img)
+            loss = loss_fn(output, batch_img)
+            batch_losses.append(loss.item())
+
+            loss.backward()
+            optimizer.step()
+
+        epoch_loss = sum(batch_losses) / len(batch_losses)
+        losses.append(epoch_loss)
+        train_log(epoch_loss, i, epochs)
```
Comment (rosinusserrano) [inline src/pml_vqvae/train.py]: I think its probably better to write a training function for each model in its own file (e.g. this function could be moved to /basline/autoencoder.py and then this file (train.py) is simply a wrapper for all of thos functions. Reason for that is that we will have different losses, especially for the VQVAE and VAE since they also have to get the latent variable Z, furthermore the pixelcnn will also not be trainable with this function.

**Review by rosinusserrano — CHANGES_REQUESTED** [review]

File: src/pml_vqvae/dataset/dataloader.py [hunk 2]
```diff
@@ -2,33 +2,12 @@
...
-        list: List of random indices
-    """
-
-    if seed:
-        rng = np.random.default_rng(seed=seed)
-    else:
-        rng = np.random.default_rng()
-
-    return rng.choice(high, n, replace=False)
+DATASET_NAMES = ["imagenet", "cifar"]
```
Comment (rosinusserrano) [inline src/pml_vqvae/dataset/dataloader.py:10]: This isnt part of this commit but so I'll just leave it here: Correct me if Im wrong, but it seems like the shuffle param in the load_data function is both applied for train and test split. I think there should be the possibility to obtain a shuffled train set and an unshuffled test split without calling that function twice

File: src/pml_vqvae/train.py (hunk 1 above)
Comment (timonpalm) [reply src/pml_vqvae/train.py]: I prefer just one train function where where we handle the logging and plotting stuff. Well, we can right the vae first into an torch.nn.module, put the encoder -> sample -> decoder logic into the forward function and the loss can be a staticmethod of that model class. 
That was at least my idea to have them all as modules.

File: src/pml_vqvae/dataset/dataloader.py (hunk 2 above)
Comment (timonpalm) [reply src/pml_vqvae/dataset/dataloader.py:10]: Yes, we can do that. But does it matter if the test set is shuffled or not. I think it makes not difference on the result.

File: src/pml_vqvae/train.py (hunk 1 above)
Comment (rosinusserrano) [reply src/pml_vqvae/train.py]: But if we have just this 1 global train function, how will we implement the copying of the gradients of the codes onto the output of the encoder, furthermore we will have to implement the "stop gradient" ($\text{sg}\[\cdot\]$) operation from the paper, which doesnt follow the usual optimizer.zero_grad() -> loss.backward() -> optimizer.step() logic. At least I think it isnt possible that way.

Still its probably useful to have a function for training the simpler models, in which case I would propose the loss_fn, optimizer and dataloaders be direct inputs into that function

File: src/pml_vqvae/dataset/dataloader.py (hunk 2 above)
Comment (rosinusserrano) [reply src/pml_vqvae/dataset/dataloader.py:10]: No it doesn't change the result, but this way we can monitor how the model changes its reconstructions/prediction of a specific sample during training. If we for example validate our model every 10 epochs, we could the save a visualization of the reconstruction of the first batch of the test_loader, which if unshuffled would be always the same.

But yeah probably not that necessary

Comment (timonpalm) [reply src/pml_vqvae/dataset/dataloader.py:10]: well, therefore is the seed
